s3prl/downstream/runner.py

```diff
@@ -390,3 +410,85 @@ def evaluate(self, split=None, logger=None, global_step=0):
             shutil.rmtree(tempdir)
 
         return [] if type(save_names) is not list else save_names
+
+    def inference(self):
+        filepath = Path(self.args.evaluate_split)
+        assert filepath.is_file()
+        filename = filepath.stem
+
+        if hasattr(self.downstream.model, "load_audio"):
+            wav = self.downstream.model.load_audio(filepath)
+        else:
+            wav, sr = torchaudio.load(str(filepath))
+            assert sr == SAMPLE_RATE
+        wavs = [wav.view(-1).to(self.args.device)]
+
+        for entry in self.all_entries:
+            entry.model.eval()
+
+        with torch.no_grad():
+            features = self.upstream.model(wavs)
+            features = self.featurizer.model(wavs, features)
+            self.downstream.model.inference(features, [filename])
+
+    def push_to_huggingface_hub(self):
+        # Setup auth
+        hf_user = os.environ.get("HF_USERNAME")
+        hf_password = os.environ.get("HF_PASSWORD")
+        huggingface_token = HfApi().login(username=hf_user, password=hf_password)
+        HfFolder.save_token(huggingface_token)
+        print(f"HF Hub user: {hf_user}")
+        
+        # Create repo on the Hub
+        upstream_model = self.args.upstream.replace("/", "__")
+        repo_name = f"superb-s3prl-{upstream_model}-{self.args.downstream}-{str(uuid.uuid4())[:8]}"
+        repo_url = HfApi().create_repo(
+            token=huggingface_token,
+            name=repo_name,
+            organization=hf_user,
+            exist_ok=True,
+            private=True,
+        )
+        print(f"Created Hub repo: {repo_url}")
+
+        # Download repo and copy templates
+        REPO_PATH = self.args.expdir + "/hub_repo/"
+        model_repo = Repository(
+            local_dir=REPO_PATH, clone_from=repo_url, use_auth_token=huggingface_token
+        )
+        TEMPLATES_PATH = Path(f"./downstream/{self.args.downstream}/hf_hub_templates/")
+        if TEMPLATES_PATH.exists():
```

Comment:
@leo19941227 I decided to allow a training run to get pushed to the Hub even when a template doesn't exist because in principle you can use the Hub to store your checkpoints for later use.

The templates currently serve two main purposes:

- Provide the necessary metadata to run SUPERB evaluations
- Provide an interface for widgets on the Hub so users can interact with the model

With the current proposal, the end-user doesn't get the above two features but can still store their results. What do you think?

Reply:
Hi @lewtun!
I also think this is a great idea! Thanks for the proposal :)